src/scenic/syntax/compiler.py

```diff
@@ -0,0 +1,1756 @@
+            args=[self.visit(hypothesis), self.visit(conclusion)],
+            keywords=[],
+        )
+
+
+def unquote(s: str) -> str:
+    if (s[:3] == s[-3:]) and s.startswith(("'''", '"""')):
+        return s[3:-3]
+    if (s[0] == s[-1]) and s.startswith(("'", '"')):
+        return s[1:-1]
+    return s
+
+
+# transformer
+
+
+class Context(IntFlag):
+    TOP_LEVEL = auto()
+    BEHAVIOR = auto()
+    MONITOR = auto()
+    COMPOSE = auto()
+    DYNAMIC = BEHAVIOR | MONITOR | COMPOSE
+
+
+class ScenicToPythonTransformer(Transformer):
+    def __init__(self, filename) -> None:
+        super().__init__(filename)
+
+        self.requirements = []
+        self.nextSyntaxId = 0
+
+        self.inBehavior: bool = False
+        "True if the transformer is processing behavior body"
+
+        self.inMonitor: bool = False
+        "True if the transformer is processing monitor body"
+
+        self.behaviorLocals: set = set()
+        "Set of variable names on the local scope of the behavior"
+
+        self.inCompose: bool = False
+        "True if the transformer is processing a `compose` block of modular scenario"
+
+        self.inTryInterrupt = False
+        self.inInterruptBlock = False
+        self.inLoop = False
+        self.usedBreak = False
+        self.usedContinue = False
+
+    @property
+    def topLevel(self):
+        return (
+            not self.inBehavior
+            and not self.inMonitor
+            and not self.inCompose
+            and not self.inTryInterrupt
+        )
+
+    def context(
+        allowedContext: Context,
+        errorBuilder: Optional[Callable[[str], str]] = None,
+    ):
+        "Mark AST node as only available inside certain contexts"
+
+        def decorator(
+            visitor: Callable[["ScenicToPythonTransformer", ast.AST], ast.AST]
+        ):
+            def check_and_visit(self: "ScenicToPythonTransformer", node: ast.AST):
+                ctx = None
+                if self.topLevel and Context.TOP_LEVEL not in allowedContext:
+                    ctx = "at the top level"
+                elif self.inBehavior and Context.BEHAVIOR not in allowedContext:
+                    ctx = "inside a behavior"
+                elif self.inMonitor and Context.MONITOR not in allowedContext:
+                    ctx = "inside a monitor"
+                elif self.inCompose and Context.COMPOSE not in allowedContext:
+                    ctx = "inside a compose block"
+                if ctx:
+                    raise self.makeSyntaxError(
+                        f'Cannot use "{node.__class__.__name__}" {ctx}'
+                        if errorBuilder is None
+                        else errorBuilder(ctx),
+                        node,
+                    )
+                return visitor(self, node)
+
+            return check_and_visit
+
+        return decorator
+
+    def generic_visit(self, node):
+        if isinstance(node, s.AST):
+            raise Exception(
```

Comment:
This is an internal error, so I would make this an assertion.

Reply:
@shumbo